### PR TITLE
FF-A Notifications

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

**Comment by b49020** [comment]
As I mentioned here [1], it would be good to unify notifications feature to use secure world donated SGI and get rid of platform specific SPI or PPI for SMC ABI too.

[1] https://lists.trustedfirmware.org/archives/list/op-tee@lists.trustedfirmware.org/message/VL7FPBS2LNT3OCZOPCSIGCPARFOMNRD7/

**Comment by etienne-lms** [comment]
> As I mentioned here [1], it would be good to unify notifications feature to use secure world donated SGI and get rid of platform specific SPI or PPI for SMC ABI too.

stm32mp platforms will continue to use a PPI over SMC ABI.

**Comment by b49020** [comment]
> stm32mp platforms will continue to use a PPI over SMC ABI.

Although I don't see ST platforms enabling notifications in upstream OP-TEE but any particular reason for this choice? I suppose there would be free SGI available there too.

**Comment by etienne-lms** [comment]
stm32mp13 does, [since not that long](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=5060e27012f7e0044ae838de59a46f563c54fb84).
As for stm32mp15, there is some legacy upstream kernel contrainst, but I still hope to overcome them, some day.

File: core/drivers/gic.c [hunk 1]
```diff
@@ -395,16 +395,18 @@ static void gic_it_raise_sgi(struct gic_data *gd __maybe_unused, size_t it,
...
-	uint64_t mask = (mask_cpu |
-			SHIFT_U64(mask_aff1, 16) |
-			SHIFT_U64(mask_id, 24)   |
-			SHIFT_U64(mask_aff2, 32) |
-			SHIFT_U64(mask_aff3, 48));
+	uint64_t mask = SHIFT_U64(mask_aff1, 16) | SHIFT_U64(mask_id, 24) |
+			SHIFT_U64(mask_aff2, 32) | SHIFT_U64(mask_aff3, 48);
+
+	if (cpu_mask & ITR_CPU_MASK_TO_THIS_CPU)
+		target_list = BIT32(mpidr & 0xf);
```
Comment (GseoC) [inline core/drivers/gic.c:657]: Shouldn't the whole bitfield of affinity 0 be covered? I see no limitations on the number of CPU per cluster

target_list = BIT32(mpidr & 0xff)?

More than 4 core per cluster is quite unusual though.

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"qemu_armv8a: enable testing of notifications using the console",
"core: separate async notification implementation",
"core: notif: remove interrupt assumptions",

`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"core: mobj_ffa.c: add reassuring comment in mobj_ffa_unregister_by_cookie()",
"qemu_armv8a: add GIC v3 redistributor base address",
"core: ffa: special treatment for FFA_ERROR",

For commit "core: gic: add gic_init_donate_sgi_to_ns()":
The commit message states "Secure world currently doesn't use any SGIs at all" which may not be true with #6440.

Commit "core: ffa: add notifications with SPMC at S-EL1" is tough, I'm not done with it, event for a shallow review :)

File: core/arch/arm/include/kernel/thread_spmc.h [hunk 2]
```diff
@@ -37,7 +40,14 @@ TEE_Result spmc_fill_partition_entry(uint32_t ffa_vers, void *buf, size_t blen,
 int spmc_read_mem_transaction(uint32_t ffa_vers, void *buf, size_t blen,
 			      struct ffa_mem_transaction_x *trans);
 
-#if !defined(CFG_CORE_SEL1_SPMC)
+#if defined(CFG_CORE_SEL1_SPMC)
+void thread_spmc_set_async_notif_intid(int intid);
+#else
+static inline void __noreturn
+thread_spmc_set_async_notif_intid(int intid __unused)
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/include/kernel/thread_spmc.h:47]: Agree, but I prefer using the same name for the argument.

File: core/arch/arm/include/optee_ffa.h [hunk 3]
```diff
@@ -109,6 +111,21 @@
...
+ * Call register usage:
+ * w3:    Service ID, OPTEE_FFA_ENABLE_ASYNC_NOTIF
+ * w4:	  Notification value to request bottom half processing
+ * w5-w7: Not used (MBZ)
+ *
+ * Return register usage:
+ * w3:    Error code, 0 on success
+ * w4-w7: Not used (MBZ)
+ */
+#define OPTEE_FFA_ENABLE_ASYNC_NOTIF	OPTEE_FFA_BLOCKING_CALL(5)
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/include/optee_ffa.h:127]: I'm not sure what happened with ID 4, but this has to be kept in synch with the kernel patches so changing this now might be more trouble than it's worth.

File: core/arch/arm/kernel/thread_spmc.c [hunk 4]
```diff
@@ -149,6 +160,26 @@ static void handle_features(struct thread_smc_args *args)
...
+		 * Looking at specific features before the Function IDs
+		 * are checked below.
+		 *
+		 * Note that mismatching values are expected to be handled
+		 * in the default case in the next switch() further down.
+		 */
+		switch (args->a1 & FFA_FEATURES_FEATURE_ID_MASK) {
+		case FFA_FEATURE_SCHEDULE_RECV_INTR:
+			if (!spmc_notif_is_ready)
+				goto err;
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/kernel/thread_spmc.c:174]: This turned out a bit more complicated than I liked. I'll refactor this function to see if you prefer that.

File: core/drivers/gic.c [hunk 5]
```diff
@@ -43,10 +43,14 @@
 #define GICD_ITARGETSR(n)	(0x800 + (n) * 4)
 #define GICD_IGROUPMODR(n)	(0xd00 + (n) * 4)
 #define GICD_SGIR		(0xF00)
+#define GICD_PIDR2		(0xFFE8)
 
 #define GICD_CTLR_ENABLEGRP0	(1 << 0)
 #define GICD_CTLR_ENABLEGRP1	(1 << 1)
 
+#define GICD_PIDR2_ARCHREV_SHIFT	(4)
+#define GICD_PIDR2_ARCHREV_MASK		(0xF)
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/drivers/gic.c:75]: I noticed that the offset is different for GICv2 (0xfe8 vs 0xffe8). I'll update.

[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/gic.c]: Thanks you to have cross checked that! I've misread both to specs as I failed to see they differ for an 'f'.
My apologies.

[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/drivers/gic.c]: No worries, it was my mistake. Thanks for reviewing.

File: core/arch/arm/kernel/thread_spmc.c [hunk 6]
```diff
@@ -1304,6 +1383,221 @@ static void handle_mem_reclaim(struct thread_smc_args *args)
...
+			notif_vm_id = vm_id;
+			notif_vm_id_valid = true;
+			ret_val = FFA_OK;
+			ret_fid = FFA_SUCCESS_32;
+		}
+
+		cpu_spin_unlock_xrestore(&spmc_notif_lock, old_itr_status);
+	}
+
+	spmc_set_args(args, ret_fid, 0, ret_val, 0, 0, 0);
```
Comment (etienne-lms) [inline core/arch/arm/kernel/thread_spmc.c:1398]: generic comment: at many places is `FFA_PARAM_MBZ` is used instead of 0.

File: core/arch/arm/kernel/thread_spmc.c [hunk 7]
```diff
@@ -1304,6 +1383,221 @@ static void handle_mem_reclaim(struct thread_smc_args *args)
...
+	uint32_t ret_fid = FFA_ERROR;
+	uint32_t old_itr_status = 0;
+
+	if (!FFA_TARGET_INFO_GET_SP_ID(args->a1) && !args->a3 && !args->a4 &&
+	    !args->a5 && !args->a6 && !args->a7) {
+		uint16_t vm_id = args->a1;
+
+		old_itr_status = cpu_spin_lock_xsave(&spmc_notif_lock);
+
+		if (!notif_async_is_started() && notif_vm_id_valid &&
```
Comment (etienne-lms) [inline core/arch/arm/kernel/thread_spmc.c:1413]: From the FF-A spec, return code should be DENIED "if the Notification bitmap is not registered or is
registered but not in a masked and non-pending state.".

File: core/arch/arm/kernel/thread_spmc.c [hunk 8]
```diff
@@ -1304,6 +1383,221 @@ static void handle_mem_reclaim(struct thread_smc_args *args)
...
+	uint32_t ret_val = FFA_INVALID_PARAMETERS;
+	uint32_t ret_fid = FFA_ERROR;
+	uint32_t old_itr_status = 0;
+	uint64_t bitmap = 0;
+	uint16_t vm_id = 0;
+
+	if (args->a5 || args->a6 || args->a7)
+		goto out;
+	if (args->a2) {
+		/* We only deal with global notifications for now. */
```
Comment (etienne-lms) [inline core/arch/arm/kernel/thread_spmc.c:1439]: nitpicking: remove period for inline comments consistency

File: core/arch/arm/kernel/thread_spmc.c [hunk 9]
```diff
@@ -1304,6 +1383,221 @@ static void handle_mem_reclaim(struct thread_smc_args *args)
...
+		ret_val = FFA_NOT_SUPPORTED;
+		goto out;
+	}
+
+	/* DST as in the destination of the eventual notification */
```
Comment (etienne-lms) [inline core/arch/arm/kernel/thread_spmc.c:1444]: "as in" ?

File: core/arch/arm/kernel/thread_spmc.c [hunk 10]
```diff
@@ -1304,6 +1383,221 @@ static void handle_mem_reclaim(struct thread_smc_args *args)
...
+		goto out;
+
+	if (OPTEE_SMC_IS_64(args->a0))
+		ret_fid = FFA_SUCCESS_64;
+	else
+		ret_fid = FFA_SUCCESS_32;
+
+	/*
+	 * Note we're only supporting physical OS kernel in normal world
+	 * with Global Notifications at that.
```
Comment (etienne-lms) [inline core/arch/arm/kernel/thread_spmc.c:1550]: "at that **moment**" ?

File: core/arch/arm/kernel/thread_spmc.c [hunk 11]
```diff
@@ -1718,6 +2033,7 @@ static TEE_Result spmc_init(void)
 	return TEE_SUCCESS;
 }
 #else /* !defined(CFG_CORE_SEL1_SPMC) */
+
```
Comment (etienne-lms) [inline core/arch/arm/kernel/thread_spmc.c:2020]: unrelated change?

File: mk/config.mk [hunk 12]
```diff
@@ -951,11 +951,17 @@ CFG_PAN ?= n
...
 $(eval $(call cfg-depends-one,CFG_PAN,CFG_ARM64_core CFG_RV64_core CFG_RV32_core))
 
+ifeq ($(filter y, $(CFG_CORE_SEL1_SPMC) $(CFG_CORE_SEL2_SPMC) \
+		  $(CFG_CORE_EL3_SPMC)),y)
+# FF-A case, handled via the FF-A ABI
+CFG_CORE_ASYNC_NOTIF ?= y
+else
 # CFG_CORE_ASYNC_NOTIF is defined by the platform to enable support
-# for sending asynchronous notifications to normal world. Note that an
+# for sending asynchronous notifications to normal world.
```
Comment (etienne-lms) [inline mk/config.mk:960]: s/interrupt/Interrupt/ below

File: core/arch/arm/kernel/thread_spmc.c (hunk 7 above)
Comment (jenswi-linaro) [reply core/arch/arm/kernel/thread_spmc.c:1413]: Thanks, I'll update.

File: core/arch/arm/kernel/thread_spmc.c (hunk 9 above)
Comment (jenswi-linaro) [reply core/arch/arm/kernel/thread_spmc.c:1444]: I'll rephrase.

File: core/arch/arm/kernel/thread_spmc.c (hunk 10 above)
Comment (jenswi-linaro) [reply core/arch/arm/kernel/thread_spmc.c:1550]: I'll rephrase it.

File: core/arch/arm/kernel/thread_spmc.c (hunk 11 above)
Comment (jenswi-linaro) [reply core/arch/arm/kernel/thread_spmc.c:2020]: Sure

**Comment by jenswi-linaro** [comment]
Addressing comments. This PR needs to be rebased. OK if I squash in the review updates and rebase?

**Comment by etienne-lms** [comment]
ok for rebasing (sorry for the delayed answer)

**Comment by jenswi-linaro** [comment]
Squashed and rebased.

**Comment by jenswi-linaro** [comment]
Tags applied and commit message of "core: gic: add gic_init_donate_sgi_to_ns()" with regards to used SGIs in the secure world.

**Comment by jenswi-linaro** [comment]
I've picked up the kernel counterpart of this, so the ABI is frozen.

**Comment by jbech-linaro** [comment]
This fails at IBART at the same way as @etienne-lms 's #6163 .
So, now we have
* Two PR's failing in the same way.
* I can easily reproduce the error locally when running with @etienne-lms branch in #6163 on two different machines.
  ``` bash
  repo init -u https://github.com/OP-TEE/manifest.git -m default.xml
  repo sync -j4
  cd build
  make toolchains -j2
  make -j28 CFG_LOCKDEP=y CFG_LOCKDEP_RECORD_STACK=n CFG_ENABLE_EMBEDDED_TESTS=n CFG_IN_TREE_EARLY_TAS=pkcs11/fd02c9da-306c-48c7-a49c-bbd827ae86ee CFG_PKCS11_TA=y BR2_PER_PACKAGE_DIRECTORIES=n CFG_EMBED_DTB=y
  ```
* When I run tip on our master, I cannot reproduce it.

**Comment by jenswi-linaro** [comment]
Rebased on master

File: core/drivers/gic.c [hunk 13]
```diff
@@ -200,33 +202,169 @@ static size_t probe_max_it(vaddr_t gicc_base __maybe_unused, vaddr_t gicd_base)
...
+		 * mask itself
+		 */
 #if defined(CFG_ARM_GICV3)
-	write_icc_pmr(0x80);
-	write_icc_igrpen1(1);
+		write_icc_pmr(0x80);
+		write_icc_igrpen1(1);
 #else
-	io_write32(gd->gicc_base + GICC_PMR, 0x80);
+		io_write32(gd->gicc_base + GICC_PMR, 0x80);
```
Comment (etienne-lms) [inline core/drivers/gic.c:320]: Before this change, when `CFG_WITH_ARM_TRUSTED_FW` is enable,
for GICv3 we called `write_icc_pmr(0x80);` + `write_icc_igrpen1(1);`
for GICv2 we called `write_icc_igrpen1(1);` + `io_write32(gd->gicc_base + GICC_CTLR, ...)`
Now its not the case. Is it on purpose?

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"core: gic: check gic version" and
"core: gic: use redistributor CPU interface" (with minor suggestion).

`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
"plat-vexpress: donate a secure SGI to normal world".

Question for commit "core: gic: add gic_init_donate_sgi_to_ns()" otherwise it looks good to me.

File: core/drivers/gic.c [hunk 14]
```diff
@@ -51,8 +51,26 @@
...
-#define GICD_CTLR_ENABLEGRP1	(1 << 1)
+#define GICD_CTLR_ENABLEGRP0	BIT32(0)
+#define GICD_CTLR_ENABLEGRP1NS	BIT32(1)
+#define GICD_CTLR_ENABLEGRP1S	BIT32(2)
+#define GICD_CTLR_ARE_S		BIT32(4)
+#define GICD_CTLR_ARE_NS	BIT32(5)
+
+/* Offsets from gic.gicr_base[core_pos] */
+#define GICR_V3_PCPUBASE_SIZE	(2 * 64 * 1024)
+#define GICR_SGIBASE_OFFSET	(64 * 1024)
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/drivers/gic.c:62]: Sure, I'll update.

File: core/drivers/gic.c (hunk 13 above)
Comment (jenswi-linaro) [reply core/drivers/gic.c:320]: Yes, that's on purpose. Should I do that in a separate commit or perhaps mention it in the commit?

Comment (jenswi-linaro) [reply core/drivers/gic.c:320]: It's actually no change in behaviour since `gic_cpu_init()` is normally not called with CFG_WITH_ARM_TRUSTED_FW=y, but with donate SGI it will be needed. I'll make a separate commit and update all the platforms.

Comment (etienne-lms) [reply core/drivers/gic.c:320]: There are many platforms that enable `CFG_WITH_ARM_TRUSTED_FW` and call `gic_cpu_init()` from `boot_secondary_init_intc()`.

Comment (jenswi-linaro) [reply core/drivers/gic.c:320]: Seems tricky to update all platforms in one go. How about deprecating `gic_cpu_init()` and add a new `gic_init_per_cpu()`? `gic_cpu_init()` is kept unchanged with regards to PMR and igrpen1/GICC_CTLR and `gic_init_per_cpu()` is expected by be by all secondary CPUs regardless of `CFG_WITH_ARM_TRUSTED_FW`. I'll submit separate PRs to tidy up the various platforms after this PR.

Comment (etienne-lms) [reply core/drivers/gic.c:320]: Sounds a good comprise. We should notify platform maintainers to update their implementation so that we can get rid of `gic_cpu_init()` afterward.

**Comment by etienne-lms** [comment]
You can rebase to solve confict and IBART test failure.

**Comment by jenswi-linaro** [comment]
Rebased to resolve the conflict. Tags applied.
I'm going to rework "core: gic: add gic_init_donate_sgi_to_ns()" as discussed above.

**Comment by jenswi-linaro** [comment]
Added "core: deprecate gic_cpu_init()" and "plat-vexpress: use gic_init_per_cpu()".

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"core: deprecate gic_cpu_init()" (with typo fixed. Feel free to discard the suggestion)
and "plat-vexpress: use gic_init_per_cpu()".

**Comment by jenswi-linaro** [comment]
Comments addressed and tag applied.

**Review by etienne-lms — COMMENTED** [review]
For commit "core: gic: add gic_init_donate_sgi_to_ns()":

Should core ensures there is no secure interrupt handler registered for donated SGIs, and counter part that secure interrupt handler cannot be registered for a donated SGI?


Typo in commit message:
"In case another non-secure SGI **is** needed secure world can donate one of its ..."

File: mk/config.mk [hunk 15]
```diff
@@ -998,11 +998,17 @@ CFG_PAN ?= n
 
 $(eval $(call cfg-depends-one,CFG_PAN,CFG_ARM64_core CFG_RV64_core CFG_RV32_core))
 
+ifeq ($(filter y, $(CFG_CORE_SEL1_SPMC) $(CFG_CORE_SEL2_SPMC) \
+		  $(CFG_CORE_EL3_SPMC)),y)
+# FF-A case, handled via the FF-A ABI
+CFG_CORE_ASYNC_NOTIF ?= y
```
Comment (etienne-lms) [inline mk/config.mk:1004]: instad of the `$(call force,...)` in _arm.mk_ and `_CFG_CORE_ASYNC_NOTIF_DEFAULT_IMPL ?= $(CFG_CORE_ASYNC_NOTIF)` in _core.mk_ directives that are a bit confusing, I would suggest here:
```patch
 ifeq ($(filter y, $(CFG_CORE_SEL1_SPMC) $(CFG_CORE_SEL2_SPMC) \
 		  $(CFG_CORE_EL3_SPMC)),y)
 # FF-A case, handled via the FF-A ABI
 CFG_CORE_ASYNC_NOTIF ?= y
+$(call force,_CFG_CORE_ASYNC_NOTIF_DEFAULT_IMPL,n)`
 else
 # CFG_CORE_ASYNC_NOTIF is defined by the platform to enable support
 # for sending asynchronous notifications to normal world.
 # Interrupt ID must be configurged by the platform too. Currently is only
 # CFG_CORE_ASYNC_NOTIF_GIC_INTID defined.
 CFG_CORE_ASYNC_NOTIF ?= n
+$(call force,_CFG_CORE_ASYNC_NOTIF_DEFAULT_IMPL,$(CFG_CORE_ASYNC_NOTIF))`
 endif
```

File: core/kernel/notif_default.c [hunk 16]
```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021-2023, Linaro Limited
+ */
+
+#include <bitstring.h>
+#include <drivers/gic.h>
```
Comment (etienne-lms) [inline core/kernel/notif_default.c:7]: replace with `#include <kernel/interrupt.h>`
add inclusion of _assert.h_ and _trace.h_.

**Review by etienne-lms — COMMENTED** [review]
Comments for commit "core: ffa: add notifications with SPMC at S-EL1".

**Comment by jenswi-linaro** [comment]
> For commit "core: gic: add gic_init_donate_sgi_to_ns()":
> 
> Should core ensures there is no secure interrupt handler registered for donated SGIs, and counter part that secure interrupt handler cannot be registered for a donated SGI?

Perhaps, but we should then also prevent registering a handler for a non-secure SGI. Can we take this later in another PR?

**Comment by jenswi-linaro** [comment]
Force push to update the commit message of "core: gic: add gic_init_donate_sgi_to_ns", the rest of the changes went into "[review]" commits.

**Comment by etienne-lms** [comment]
> Perhaps, but we should then also prevent registering a handler for a non-secure SGI. Can we take this later in another PR?

Sure.

Build errors:
```
 core/kernel/notif_default.c:8:10: fatal error: drivers/interrupt.h: No such file or directory
    8 | #include <drivers/interrupt.h>
      |          ^~~~~~~~~~~~~~~~~~~~~
```
s/`drivers/interrupt.h`/`kernel/interrupt.h`/
Maybe should we move kernel/interrupt.h_ to _drivers/interrupt.h_?

File: core/drivers/gic.c [hunk 17]
```diff
@@ -260,11 +260,11 @@ static void gic_legacy_sync_sgi_config(struct gic_data *gd)
 		/* Ignore matching bits */
 		if (!(BIT32(n) & (grp0 ^ gd->per_cpu_group_status)))
 			continue;
-		need_sync = true;
 		/*
 		 * SGI-n differs from primary CPU configuration,
 		 * let's sync up.
 		 */
+		need_sync = true;
```
Comment (etienne-lms) [inline core/drivers/gic.c:267]: I forgot to mention my comment also applied to _gicv3_sync_sgi_config_ line 225.

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
"core: gic: add gic_init_donate_sgi_to_ns()" **(edited: with comment addressed :)**

`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
"core: ffa: add notifications with SPMC at S-EL1".

File: core/drivers/gic.c (hunk 17 above)
Comment (jenswi-linaro) [reply core/drivers/gic.c:267]: I'll fix it.

**Comment by jenswi-linaro** [comment]
I'll squash and apply tags once this has passed CI.

**Comment by jenswi-linaro** [comment]
Squashed and tags applied. All commits are reviewed and should be ready to merge.

File: core/include/drivers/gic.h [hunk 18]
```diff
@@ -30,7 +30,12 @@
  */
 
 /* Initialize GIC */
-void gic_init(paddr_t gicc_base_pa, paddr_t gicd_base_pa);
+void gic_init_v3(paddr_t gicc_base_pa, paddr_t gicd_base_pa,
+		 paddr_t gicr_base_pa);
+static inline void gic_init(paddr_t gicc_base_pa, paddr_t gicd_base_pa)
+{
+	gic_init_v3(gicc_base_pa, gicd_base_pa, 0);
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/include/drivers/gic.h:45]: I'm sorry, I missed that. Would it work to only check for the redistributor address if affinity routing is enabled? Or should we rather let it be optional and only panic or fail if a function that needs it is called?

[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/include/drivers/gic.h:45]: @glneo, will this work? https://github.com/OP-TEE/optee_os/pull/6516
